### PR TITLE
enabling IPCC module of the stm32WB55

### PR DESCRIPTION
**Pull request by FRASTM**
This PR is the introduction  of the BLE for the stm32WB55 

Requires https://github.com/stm32duino/STM32duinoBLE/pull/17

### CODE REVIEW

File: variants/PNUCLEO_WB55RG/variant.cpp [hunk 1]
```diff
@@ -103,6 +103,33 @@ WEAK void SystemClock_Config(void)
   __HAL_RCC_LSEDRIVE_CONFIG(RCC_LSEDRIVE_LOW);
   /* Configure the main internal regulator output voltage */
   __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE1);
+#if  defined (ARDUINO_PNUCLEO_WB55RG)
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/variant.cpp:106]: You don't need to use this define, it is always true as it is in the variant.

File: variants/PNUCLEO_WB55RG/otp.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  #define PACKED_STRUCT struct PACKED__
+#elif defined (__ICCARM__)
+  #define PACKED_STRUCT __packed struct
+#elif
+  #define PACKED_STRUCT __packed struct
+#endif
+
+  /* Exported types ------------------------------------------------------------*/
+  typedef  PACKED_STRUCT
+  {
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/otp.h:55]: No need to define this as Arduino only manage one toolchain.
Moreover the packed attribute does not need to be redefined. It is already managed here:
https://github.com/stm32duino/Arduino_Core_STM32/blob/bee856121e7406314226506d20617b523a092a70/system/Drivers/STM32WBxx_HAL_Driver/Inc/stm32wbxx_hal_def.h#L115

```suggestion
  typedef struct __packed {
```

File: variants/PNUCLEO_WB55RG/otp.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    uint8_t   bd_address[6];
+    uint8_t   hse_tuning;
+    uint8_t   id;
+  } OTP_ID0_t;
+
+  /* Exported constants --------------------------------------------------------*/
+  /* External variables --------------------------------------------------------*/
+  /* Exported macros -----------------------------------------------------------*/
+  /* Exported functions ------------------------------------------------------- */
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/otp.h:64]:
```suggestion
```

File: variants/PNUCLEO_WB55RG/otp.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#define CFG_OTP_BASE_ADDRESS    OTP_AREA_BASE
+
+#define CFG_OTP_END_ADDRESS      OTP_AREA_END_ADDR
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/otp.h:67]:
```suggestion
#define CFG_OTP_END_ADDRESS     OTP_AREA_END_ADDR
```

File: variants/PNUCLEO_WB55RG/variant.cpp [hunk 5]
```diff
@@ -88,71 +89,105 @@ const PinName digitalPin[] = {
...
-  PeriphClkInitStruct.SmpsDivSelection = RCC_SMPSCLKDIV_RANGE0;
-
-  if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInitStruct) != HAL_OK) {
-    Error_Handler();
-  }
-  /* Enable MSI Auto calibration */
-  HAL_RCCEx_EnableMSIPLLMode();
+    RCC_OscInitTypeDef RCC_OscInitStruct = {0};
+    RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};
+    RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {0};
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/variant.cpp:135]: Avoid **_-Wmissing-field-initializers_** warning
```suggestion
    RCC_OscInitTypeDef RCC_OscInitStruct = {};
    RCC_ClkInitTypeDef RCC_ClkInitStruct = {};
    RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {};
```

File: variants/PNUCLEO_WB55RG/variant.cpp [hunk 6]
```diff
@@ -88,71 +89,105 @@ const PinName digitalPin[] = {
...
+
+    /**
+     * Read HSE_Tuning from OTP
+     */
+    p_otp = (OTP_ID0_t *) OTP_Read(0);
+    if (p_otp) {
+        LL_RCC_HSE_SetCapacitorTuning(p_otp->hse_tuning);
+    }
+
+    return;
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/variant.cpp:123]:
```suggestion
```

File: variants/PNUCLEO_WB55RG/variant.cpp [hunk 7]
```diff
@@ -88,71 +89,105 @@ const PinName digitalPin[] = {
...
+    RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};
+    RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {0};
+
+    Config_HSE();
+
+    __HAL_RCC_LSEDRIVE_CONFIG(RCC_LSEDRIVE_LOW);
+    __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE1);
+
+    /* This prevents the CPU2 (M0+) to disable the HSI48 oscillator
+    while (LL_HSEM_1StepLock(HSEM, CFG_HW_CLK48_CONFIG_SEMID)); */
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/variant.cpp:143]: Remove if not needed
```suggestion
```

File: variants/PNUCLEO_WB55RG/variant.cpp [hunk 8]
```diff
@@ -88,71 +89,105 @@ const PinName digitalPin[] = {
...
+    RCC_OscInitStruct.HSI48State = RCC_HSI48_ON;
+    RCC_OscInitStruct.LSEState = RCC_LSE_ON;
+    RCC_OscInitStruct.HSICalibrationValue = RCC_HSICALIBRATION_DEFAULT;
+    RCC_OscInitStruct.PLL.PLLState = RCC_PLL_NONE;
+    if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
+        Error_Handler();
+    }
+
+    /** Configure the SYSCLKSource, HCLK, PCLK1 and PCLK2 clocks dividers
+    */
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/variant.cpp:158]:
```suggestion
    /* Configure the SYSCLKSource, HCLK, PCLK1 and PCLK2 clocks dividers */
```

File: variants/PNUCLEO_WB55RG/variant.h [hunk 9]
```diff
@@ -120,6 +120,9 @@ extern "C" {
 // for EEPROM emulation to the last 512k pages.
 #define FLASH_PAGE_NUMBER       127
 
+
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/variant.h:123]: Remove extra line
```suggestion
```

**Review by fpistm — CHANGES_REQUESTED** [review]
In addition to my review comments, please:
- Rebase and squash on top of the master and change commit message to something like:

> [PNUCLEO_WB55RG] BLE support configuration
> 
> - Enable IPCC HAL module
> - Manage OTP to ...
> - Update clock configuration
> 

- Apply astyle (script available in CI/astyle/astyle.py)

**Comment by FRASTM** [comment]
fix according to the review (mainly astyle)
squash in one commit

File: variants/PNUCLEO_WB55RG/otp.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+extern "C" {
+#endif
+
+/* Includes ------------------------------------------------------------------*/
+/* ----------------------------------- *
+ *  Packed usage (compiler dependent)  *
+ * ----------------------------------- */
+#undef PACKED__
+#undef PACKED_STRUCT
+
```
Comment (fpistm) [inline variants/PNUCLEO_WB55RG/otp.h:35]: Useless can be removed

```suggestion
/* Exported types ------------------------------------------------------------*/
```

**Review by fpistm — CHANGES_REQUESTED** [review]

**Review by fpistm — APPROVED** [review]
LGTM
Thanks @FRASTM

**Pull request merged**